import axios, { CanceledError } from "axios";
import { useEffect, useState } from "react";

interface User {
	id: number;
	name: string;
}
function App() {
	const [users, setUsers] = useState<User[]>([]);
	const [err, setErr] = useState("");
	const [isLoading, setLoading] = useState(false);

	useEffect(() => {
		const controller = new AbortController();
		setLoading(true);
		axios
			.get<User[]>("https://jsonplaceholder.typicode.com/users", {
				signal: controller.signal,
			})
			.then((res) => {
				setUsers(res.data);
				setLoading(false);
			})
			.catch((err) => {
				if (err instanceof CanceledError) return; // need to import
				setErr(err.message);
				setLoading(false);
			});
		// .finally(() => setLoading(false)); // Doesn't Work on Strike Mode

		return () => controller.abort();
	}, []);

	return (
		<div className="container">
			{err && <p className="text-danger">{err}</p>}
			{isLoading && <div className="spinner-border"></div>}
			<ul>
				{users.map((user) => (
					<li key={user.id}>{user.name}</li>
				))}
			</ul>
		</div>
	);
}

export default App;
